Extend RunWatchSemantics test by a new flavour and fill the gap

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:
the new scenario proves that a storage layer follows RV even when sendInitialEvents flag was set to false.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

p0lyn0mial commented:
/assign @wojtek-t

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go:
```diff
@@ -1319,6 +1323,17 @@ func RunWatchSemantics(ctx context.Context, t *testing.T, store storage.Interfac
 			},
 		},
+		{
+			name:                               "sendInitialEvents=false, RV=useCurrentRV",
```
Can you additionally add a similar case (useCurrentRV) for 
a) sendInitialEvents=true 
b) sendInitialEvents=true & allwWatchBookmarks=true

[the one you added looks good already]

p0lyn0mial replied:
Actually I started wondering if we need the case I added. 
How does it differ from 
 - `sendInitialEvents=false, RV=1` 
- or the one you proposed `b` how does it differ from `sendInitialEvents=true & allowWatchBookmarks=true, RV=1`.

In both cases `RV=1` and `RV=useCurrentRV` are behind "the initial events". 

I think the `useCurrentRV` could be read after adding the initial events and that would give us a new flavour. WDYT?

I have prepared the following matrix (I will add it to the source code) for better visualising the current state and potential gaps in testing.

```
// RunWatchSemantics test the following cases:
//
// +-----------------+---------------------+-------------------+
// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents |
// +-----------------+---------------------+-------------------+
// | Unset           | -                   | -/-               |
// | Unset           | false               | true/false        |
// | 0               | true                | true/-            |
// | 0               | false               | true/false        |
// | 1               | true                | true/-            |
// | 1               | false               | true/-            |
// +-----------------+---------------------+-------------------+
//
// where:
// - false indicates the value of the param was set to "false" by a test case
// - true  indicates the value of the param was set to "true" by a test case
```


The first case is interesting and I'm surprised we have missed it. But then I realised that we have `RunWatchSemanticInitialEventsExtended` which actually covers `allowWatchBookmarks=true, sendInitialEvents=true, RV=unset`. We moved it to a separate test because it ensure the global rv will be delivered as part of the initial bookmarks. Which is fine but I think we could add a standard case here as well.

wojtek-t replied:
> I think the useCurrentRV could be read after adding the initial events and that would give us a new flavour.

Agree.

p0lyn0mial replied:
Are you also OK with adding more cases to fill the matrix ?

p0lyn0mial replied:
pushed, PTAL.

jiahuif commented:
/triage accepted

p0lyn0mial committed:
storage/testing/watcher_tests: add new flavour to RunWatchSemantics test and fill the gap

the new flavour proves that a storage layer follows RV
which is equal to the global rv

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go:
```diff
@@ -1222,6 +1222,24 @@ func RunSendInitialEventsBackwardCompatibility(ctx context.Context, t *testing.T
+// +-----------------+---------------------+-------------------+
+// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents |
+// +=================+=====================+===================+
+// | Unset           | true                | true/false        |
```
now, i'm fine with merging the rows based on the `RV` value.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go:
```diff
@@ -1277,6 +1305,14 @@ func RunWatchSemantics(ctx context.Context, t *testing.T, store storage.Interfac
+			sendInitialEvents:                  &falseVal,
+			resourceVersion:                    "1",
+			initialPods:                        []*example.Pod{makePod("1"), makePod("2")},
+			expectedInitialEventsInRandomOrder: addEventsFromCreatedPods,
```
Why?

For sendInitialEvents=false, we don't expect initial events ?

What we rather expect is the actual stream of events (so should be strict order).

p0lyn0mial replied:
no, I don't think so, we expect events > `1` (this is different than initial events) which are taken from the map so the order is not guaranteed.

wojtek-t replied:
We're not taking from the map - we're only taking the state from the map.
Here we're actually taking a watch stream.

wojtek-t replied:
watch stream = transaction log

p0lyn0mial replied:
ah, now i see you what you mean, we actually have "two maps", I will double check that

p0lyn0mial replied:
okay, you were right, thanks!

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go:
```diff
@@ -1285,6 +1321,28 @@ func RunWatchSemantics(ctx context.Context, t *testing.T, store storage.Interfac
+			name:                                 "allowWatchBookmarks=true, sendInitialEvents=false, RV=unset",
+			allowWatchBookmarks:                  true,
+			sendInitialEvents:                    &falseVal,
+			initialPods:                          []*example.Pod{makePod("1")},
```
Also - let's unify the test cases:
- let's always [in every single case] create 2 initial pods [or maybe even 3?]
- let's always create 2 pods after establishing a watch

p0lyn0mial replied:
yes, this is a good idea, let's do that.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go:
```diff
@@ -1266,6 +1287,13 @@ func RunWatchSemantics(ctx context.Context, t *testing.T, store storage.Interfac
 				return []watch.Event{initialEventsEndFromLastCreatedPod(createdInitialPods)}
 			},
 		},
+		{
```
i was thinking about it but didn't want to introduce a bigger diff, will reorder :)

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go:
```diff
@@ -1300,6 +1358,13 @@ func RunWatchSemantics(ctx context.Context, t *testing.T, store storage.Interfac
+			sendInitialEvents:                  &falseVal,
+			resourceVersion:                    "1",
+			initialPods:                        []*example.Pod{makePod("1"), makePod("2")},
+			expectedInitialEventsInRandomOrder: addEventsFromCreatedPods,
```
> In general the rule is - the "randomOrder" may be present only in cases where sendINitialEvents is true.

not necessarily, also in all cases that provide an RV that <= what the cache thinks

wojtek-t commented:
This is great now!

/lgtm
/approve
/retest

p0lyn0mial commented:
/test pull-kubernetes-e2e-gce
